File: solutions/typescript/resistor-color-trio/1/resistor-color-trio.ts
const COLOR_MAP = {
  "black": 0,
  "brown": 1,
  "red": 2,
  "orange": 3,
  "yellow": 4,
  "green": 5,
  "blue": 6,
  "violet": 7,
  "grey": 8,
  "white": 9,
};

const UNITS = ['ohms', 'kiloohms', 'megaohms', 'gigaohms'];

export function decodedResistorValue(bands: (keyof typeof COLOR_MAP)[]): string {
  let value = COLOR_MAP[bands[0]] * 10 + COLOR_MAP[bands[1]];
  let pow = COLOR_MAP[bands[2]];
  if (value === 0) {
    pow = 0;
  } else if (value % 10 === 0) {
    pow += 1;
    value /= 10;
  }
  let unit = Math.floor(pow / 3);
  value *= Math.pow(10, (pow % 3));
  return `${value} ${UNITS[unit]}`;
}

File: solutions/c/resistor-color-trio/1/resistor_color_trio.c
#include "resistor_color_trio.h"

#include <math.h>

resistor_value_t color_code(const resistor_band_t *bands) {
    resistor_value_t actual;
    actual.value = (10 * bands[0]) + bands[1];

    if (actual.value == 0) {
        actual.unit = OHMS;
    } else {
        short power = bands[2];
        if (actual.value % 10 == 0) {
            actual.value /= 10;
            power += 1;
        }
        actual.unit = power / 3; // 2 -> 0, 4 -> 1, ...
        actual.value *= (short)pow(10, (power % 3));
    }
    return actual;
}